BlueBrain/CoreNeuron
# Support for shared libraries in GPU execution (python launch support)

**Description**


**Summary**

* mod2c now generates code without need of global variables
* coreneuron and mechanism library can be built as shared and it
  enables launching coreneuron on GPU via pyton
* scopmath library can be also shared
* removed acc/openmp global annotations for celsius, pi and secondorder
  and they don't need to be copied on GPU

**TODOs**

- [x] MOD2C generates code without using globals / acc declare
      See see BlueBrain/mod2c/pull/78
- [x] Basic test with special and python on GPU
      See https://github.com/BlueBrain/CoreNeuron/issues/141#issuecomment-1086742194
- [x] Link issues with CUDA part e.g. nrnran123.cu functions result
      into link errors, see
      https://github.com/BlueBrain/CoreNeuron/issues/141#issuecomment-1086742194
      @olupton to rescue here!
- [x] Check celsius usage within coreneuron source code
- [x] Investigate why acc_deviceptr(ml->data) returns host
      pointer when coreneuron is launched via python. See
      https://github.com/BlueBrain/CoreNeuron/issues/141#issuecomment-1086746848
- [x] Run neuron test suite and external models like olfactory-bulb via
      python
- [x] Update submodule BlueBrain/mod2c/pull/78
- [x] Update submodule https://github.com/BlueBrain/nmodl/pull/904
- [x] Investigate / create follow up issue because OpenACC (offload) and OpenMP (host threading) seem to interact badly when launched from Python. Typical error: `call to cuLaunchKernel returned error 400: Invalid handle`
- [x] Cleanup / create follow up issue around unified memory mode -- this needs to be "all or nothing", and at present its somewhere in the middle... --> https://github.com/BlueBrain/CoreNeuron/issues/594

Closes #141.
Closes https://github.com/BlueBrain/CoreNeuron/issues/599.

**How to test this?**

Build neuron master with this PR branch of coreneuron including updated mod2c submodule.

```bash
module load unstable gcc nvhpc cuda hpe-mpi cmake  python-dev

cmake .. -DCMAKE_INSTALL_PREFIX=`pwd`/install -DCORENRN_ENABLE_GPU=ON -DCORENRN_ENABLE_NMODL=OFF -DCORENRN_ENABLE_MPI=ON  -DNRN_ENABLE_CORENEURON=ON -DNRN_ENABLE_INTERVIEWS=OFF -DNRN_ENABLE_TESTS=OFF
make -j12
make install
```

No compile `ringtest` with `nrnivmodl -coreneuron` and launch ringtest on GPU using `python` as well as `special`. See https://github.com/BlueBrain/CoreNeuron/issues/141#issuecomment-1086742194 

**Test System**

 - Compiler: NVHPC & CUDA on BB5
 - Version: this PR and neuron@master
 - Backend: GPU

CI_BRANCHES:NEURON_BRANCH=olupton/coreneuron-gpu-dynamic-loading,SPACK_BRANCH=olupton/coreneuron-gpu-dynamic
